### ISSUE TITLE
[Phase 5] Model Implementation

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/real-estate-valuation-engine#7]
## Summary

Implement the core valuation models: Gaussian Process for spatial correlation, XGBoost with monotonic constraints, and Quantile Forest for uncertainty quantification. Combine them in an ensemble.

**Part of:** #1
**Depends on:** #6 (Phase 4)

## Objectives

- [ ] Implement `GaussianProcessModel` with RBF kernel
- [ ] Implement `MonotonicXGBoostModel` with constraints
- [ ] Implement `QuantileForestModel` with prediction intervals
- [ ] Implement `EnsembleValuationModel` combining all three
- [ ] Create consistent `PredictionResult` format
- [ ] Write unit tests for each model

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/models/gaussian_process.py` | NEW | Create | GP model implementation |
| `src/models/xgboost_model.py` | NEW | Create | XGBoost with monotonic |
| `src/models/quantile_forest.py` | NEW | Create | QRF implementation |
| `src/models/ensemble.py` | 1-80 | Implement | Ensemble model |
| `tests/models/test_*.py` | NEW | Create | Unit tests |

### GaussianProcessModel

```python
# src/models/gaussian_process.py
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import RBF, WhiteKernel
from .base import BaseModel, PredictionResult

class GaussianProcessModel(BaseModel):
    """GP for capturing spatial correlation."""

    def __init__(self, length_scale=0.01, noise_level=0.1):
        self.kernel = RBF(length_scale=length_scale) + WhiteKernel(noise_level=noise_level)
        self._model = None

    def fit(self, X, y):
        self._model = GaussianProcessRegressor(
            kernel=self.kernel,
            normalize_y=True,
            random_state=42
        )
        self._model.fit(X, y)
        return self

    def predict(self, X):
        mean, std = self._model.predict(X, return_std=True)
        return PredictionResult(
            point_estimate=mean,
            lower_bound=mean - 1.96 * std,
            upper_bound=mean + 1.96 * std
        )

    @property
    def feature_importances_(self):
        return None  # GP doesn't provide feature importances
```

### MonotonicXGBoostModel

```python
# src/models/xgboost_model.py
import xgboost as xgb
from .base import BaseModel, PredictionResult

class MonotonicXGBoostModel(BaseModel):
    """XGBoost with monotonic constraints for logical relationships."""

    POSITIVE_MONOTONIC = ['sqft', 'lot_size', 'bedrooms', 'bathrooms', 'year_built']

    def __init__(self, n_estimators=100, max_depth=6, learning_rate=0.1):
        self.n_estimators = n_estimators
        self.max_depth = max_depth
        self.learning_rate = learning_rate
        self._model = None
        self._feature_names = None

    def fit(self, X, y):
        self._feature_names = list(X.columns)

        # Build monotonic constraints
        constraints = [1 if col in self.POSITIVE_MONOTONIC else 0
                      for col in self._feature_names]

        self._model = xgb.XGBRegressor(
            n_estimators=self.n_estimators,
            max_depth=self.max_depth,
            learning_rate=self.learning_rate,
            monotone_constraints=tuple(constraints),
            random_state=42
        )
        self._model.fit(X, y)
        return self

    def predict(self, X):
        return PredictionResult(point_estimate=self._model.predict(X))

    @property
    def feature_importances_(self):
        return self._model.feature_importances_ if self._model else None
```

### QuantileForestModel

```python
# src/models/quantile_forest.py
from quantile_forest import RandomForestQuantileRegressor
from .base import BaseModel, PredictionResult

class QuantileForestModel(BaseModel):
    """Random Forest with quantile predictions for uncertainty."""

    def __init__(self, n_estimators=100, quantiles=(0.05, 0.5, 0.95)):
        self.n_estimators = n_estimators
        self.quantiles = quantiles
        self._model = None

    def fit(self, X, y):
        self._model = RandomForestQuantileRegressor(
            n_estimators=self.n_estimators,
            random_state=42
        )
        self._model.fit(X, y)
        return self

    def predict(self, X):
        preds = self._model.predict(X, quantiles=self.quantiles)
        return PredictionResult(
            point_estimate=preds[:, 1],  # Median
            lower_bound=preds[:, 0],     # 5th percentile
            upper_bound=preds[:, 2],     # 95th percentile
            confidence_level=0.90
        )

    @property
    def feature_importances_(self):
        return self._model.feature_importances_ if self._model else None
```

### Test: Monotonic Constraints

```python
# tests/models/test_xgboost.py
import pytest
import numpy as np
import pandas as pd
from src.models.xgboost_model import MonotonicXGBoostModel

def test_monotonic_constraint_sqft():
    """Larger sqft should never predict lower price."""
    np.random.seed(42)
    n = 100

    X = pd.DataFrame({
        'sqft': np.linspace(1000, 2000, n),
        'bedrooms': np.random.randint(2, 5, n),
        'other': np.random.randn(n)
    })
    y = X['sqft'] * 200 + np.random.randn(n) * 10000

    model = MonotonicXGBoostModel()
    model.fit(X, y)

    # Test predictions with increasing sqft only
    X_test = pd.DataFrame({
        'sqft': [1000, 1500, 2000],
        'bedrooms': [3, 3, 3],
        'other': [0, 0, 0]
    })

    preds = model.predict(X_test).point_estimate

    # Predictions should be monotonically increasing
    assert preds[1] >= preds[0], "1500 sqft should predict >= 1000 sqft"
    assert preds[2] >= preds[1], "2000 sqft should predict >= 1500 sqft"
```

## Verification

```bash
pytest tests/models/ -v
```

## Technical Challenges

1. **GP memory** - O(n^3) complexity limits to ~5,000 samples
   - Solution: Subsample or use regional sub-models
2. **Monotonic constraint format** - XGBoost uses tuple, order matters
   - Solution: Build dynamically from feature names
3. **Quantile prediction format** - Different from point predictions
   - Solution: Unified PredictionResult dataclass

## Definition of Done

- [ ] Each model trains without errors
- [ ] Predictions are within reasonable ranges
- [ ] Monotonic constraints verified
- [ ] Prediction intervals contain most actual values (~90%)
- [ ] All unit tests pass

## Labels

`phase-5`, `models`, `ml`, `enhancement`
